chain/store/snapshot.go

```diff
@@ -68,6 +101,300 @@ func (cs *ChainStore) Import(ctx context.Context, r io.Reader) (*types.TipSet, e
+	Message
+	Receipt
+	State
+	Dag
+)
+
+func (s *walkScheduler) work(ctx context.Context, todo *walkTask, results chan *walkResult) error {
+	defer s.taskWg.Done()
+	// unseen cid, its a result
+	results <- &walkResult{c: todo.c}
+
+	// extract relevant dags to walk from the block
+	if todo.taskType == 0 {
+		blk := todo.c
+		data, err := s.store.Get(ctx, blk)
+		if err != nil {
+			return err
+		}
+		var b types.BlockHeader
+		if err := b.UnmarshalCBOR(bytes.NewBuffer(data.RawData())); err != nil {
+			return xerrors.Errorf("unmarshaling block header (cid=%s): %w", blk, err)
+		}
+		if b.Height%1000 == 0 {
+			log.Infow("block export", "height", b.Height)
+		}
+		if b.Height == 0 {
+			log.Infow("GENESIS")
+			for i := range b.Parents {
+				s.enqueueIfNew(&walkTask{
+					c:        b.Parents[i],
+					taskType: Dag,
+				})
+			}
+			s.enqueueIfNew(&walkTask{
+				c:        b.ParentStateRoot,
+				taskType: State,
+			})
+			return nil
+		}
+		// enqueue block parents
+		for i := range b.Parents {
+			s.enqueueIfNew(&walkTask{
+				c:        b.Parents[i],
+				taskType: Block,
+			})
+		}
+		if s.cfg.tail.Height() >= b.Height {
+			log.Debugw("tail reached", "cid", blk.String())
+			return nil
+		}
+
+		if s.cfg.includeMessages {
+			// enqueue block messages
+			s.enqueueIfNew(&walkTask{
+				c:        b.Messages,
+				taskType: Message,
+			})
+		}
+		if s.cfg.includeReceipts {
+			// enqueue block receipts
+			s.enqueueIfNew(&walkTask{
+				c:        b.ParentMessageReceipts,
+				taskType: Receipt,
+			})
+		}
+		if s.cfg.includeState {
+			s.enqueueIfNew(&walkTask{
+				c:        b.ParentStateRoot,
+				taskType: State,
+			})
+		}
+
+		return nil
+	}
+	data, err := s.store.Get(ctx, todo.c)
+	if err != nil {
+		return err
+	}
+	return cbg.ScanForLinks(bytes.NewReader(data.RawData()), func(c cid.Cid) {
+		if todo.c.Prefix().Codec != cid.DagCBOR || todo.c.Prefix().MhType == mh.IDENTITY {
+			return
+		}
+
+		s.enqueueIfNew(&walkTask{
+			c:        c,
+			taskType: Dag,
+		})
+	})
+}
+
+func (cs *ChainStore) WalkSnapshotRange(ctx context.Context, store bstore.Blockstore, head, tail *types.TipSet, messages, receipts, stateroots bool, workers int64, cb func(cid.Cid) error) error {
+	start := time.Now()
+	log.Infow("walking snapshot range", "head", head.Key(), "tail", tail.Key(), "messages", messages, "receipts", receipts, "stateroots", stateroots, "workers", workers, "start", start)
+	tasks := []*walkTask{}
+	for i := range head.Blocks() {
+		tasks = append(tasks, &walkTask{
+			c:        head.Blocks()[i].Cid(),
+			taskType: 0,
+		})
+	}
+
+	cfg := &walkSchedulerConfig{
+		numWorkers:      workers,
+		tail:            tail,
+		includeMessages: messages,
+		includeState:    stateroots,
+		includeReceipts: receipts,
+	}
+
+	pw, ctx := newWalkScheduler(ctx, store, cfg, tasks...)
+	results := make(chan *walkResult)
+	pw.startScheduler(ctx)
+	pw.startWorkers(ctx, results)
+
+	done := make(chan struct{})
+	var cbErr error
+	go func() {
```

Comment:
potential goroutine leak if things error somewhere, as close(results) is down below.

Reply:
results is now always closed: https://github.com/filecoin-project/lotus/pull/9192/files#diff-740d4bd1f10c65674bc1d0a54f1fc4192d8008b0c08179a300b42116ea0454a9R408